src/main/webapp/app/localvc/commit-history/commit-history.component.ts

```diff
@@ -0,0 +1,106 @@
+})
+export class CommitHistoryComponent implements OnInit, OnDestroy {
+    readonly PROGRAMMING = ExerciseType.PROGRAMMING;
+    readonly dayjs = dayjs;
+
+    studentParticipation: StudentParticipation;
+    participationId: number;
+    paramSub: Subscription;
+    commits: CommitInfo[];
+    commitsInfoSubscription: Subscription;
+    participationSub: Subscription;
+
+    constructor(
+        private route: ActivatedRoute,
+        private programmingExerciseParticipationService: ProgrammingExerciseParticipationService,
+    ) {}
+
+    ngOnDestroy() {
+        this.paramSub?.unsubscribe();
+        this.commitsInfoSubscription?.unsubscribe();
+        this.participationSub?.unsubscribe();
+    }
+
+    /**
+     * On init, subscribe to the route params to get the participation id and load the participation.
+     */
+    ngOnInit() {
+        this.paramSub = this.route.params.subscribe((params) => {
+            this.participationId = Number(params['participationId']);
+            this.loadParticipation();
+        });
+    }
+
+    /**
+     * Load the participation with all results. Calls the handleCommits method after the participation is loaded.
+     * @private
+     */
+    private loadParticipation() {
+        this.participationSub = this.programmingExerciseParticipationService
+            .getStudentParticipationWithAllResults(this.participationId)
+            .pipe(
+                tap((participation) => {
+                    this.studentParticipation = participation;
+                    this.studentParticipation.results?.forEach((result) => {
+                        result.participation = participation!;
+                    });
+                }),
+            )
+            .subscribe({
+                next: () => {
+                    this.handleCommits();
+                },
+            });
+    }
+
+    /**
+     * Retrieves the commit history for the participation and filters out the commits that have no submission.
+     * The last commit is always the template commit and is added to the list of commits.
+     * @private
+     */
+    private handleCommits() {
+        this.commitsInfoSubscription = this.programmingExerciseParticipationService.retrieveCommitHistoryForParticipation(this.participationId).subscribe((commits) => {
+            this.commits = this.sortCommitsByTimestampDesc(commits);
+            this.setCommitResults();
+        });
+    }
+
+    /**
+     * Sets the result of the commit if it exists.
+     * @private
+     */
+    private setCommitResults() {
+        this.commits.forEach((commit) => {
+            this.studentParticipation.results?.forEach((result) => {
+                const submission = result.submission as ProgrammingSubmission;
+                if (submission) {
+                    if (submission.commitHash === commit.hash) {
+                        commit.result = result;
+                    }
+                }
+            });
+        });
```

Comment:
Please check if this is correct. One submission can have multiple results.

I think it would be better if the server would return participations with submissions and their latest result, then you don't need do iterate through all results but just find ht submission with a matching commit hash.

Reply:
As the current solution works and only causes a slight performance difference I'd suggest to implement it how you said in a follow up as quite a bit of refactoring and rewriting tests is necessary